File: NinePSharp.Messages.FSharp/messages/Txattrcreate.fs
namespace NinePSharp.Messages.FSharp

open System
open System.Buffers.Binary
open NinePSharp.Constants
open NinePSharp.Interfaces
open NinePSharp.Messages
open NinePSharp.Messages.FSharp.Protocol

type Txattrcreate =
    { Size: uint32
      Tag: uint16
      Fid: uint32
      Name: string
      AttrSize: uint64
      Flags: uint32 }

    interface ISerializable with
        member this.Size = this.Size
        member this.Type = MessageTypes.Txattrcreate
        member this.Tag = this.Tag
        member this.WriteTo(span: Span<byte>) =
            let mutable offset = 0
            writeHeaders this.Size this.Tag MessageTypes.Txattrcreate span &offset
            BinaryPrimitives.WriteUInt32LittleEndian(span.Slice(offset, 4), this.Fid); offset <- offset + 4
            writeString this.Name span &offset
            BinaryPrimitives.WriteUInt64LittleEndian(span.Slice(offset, 8), this.AttrSize); offset <- offset + 8
            BinaryPrimitives.WriteUInt32LittleEndian(span.Slice(offset, 4), this.Flags)

    static member Parse (data: ReadOnlySpan<byte>) =
        let mutable offset = 0
        let size, tag = readHeaders data &offset
        let fid = BinaryPrimitives.ReadUInt32LittleEndian(data.Slice(offset, 4))
        offset <- offset + 4
        let name = readString data &offset
        let attrSize = BinaryPrimitives.ReadUInt64LittleEndian(data.Slice(offset, 8))
        offset <- offset + 8
        let flags = BinaryPrimitives.ReadUInt32LittleEndian(data.Slice(offset, 4))
        { Size = size; Tag = tag; Fid = fid; Name = name; AttrSize = attrSize; Flags = flags }

File: NinePSharp/messages/Txattrcreate.cs
namespace NinePSharp.Messages;

using System;
using System.Buffers.Binary;
using NinePSharp.Constants;
using NinePSharp.Interfaces;
using NinePSharp.Protocol;

public readonly struct Txattrcreate : ISerializable
{
    public uint Size { get; }
    public MessageTypes Type => MessageTypes.Txattrcreate;
    public ushort Tag { get; }
    public uint Fid { get; }
    public string Name { get; }
    public ulong AttrSize { get; }
    public uint Flags { get; }

    public Txattrcreate(uint size, ushort tag, uint fid, string name, ulong attrSize, uint flags)
    {
        Size = size;
        Tag = tag;
        Fid = fid;
        Name = name;
        AttrSize = attrSize;
        Flags = flags;
    }

    public Txattrcreate(ReadOnlySpan<byte> data)
    {
        Size = BinaryPrimitives.ReadUInt32LittleEndian(data[..4]);
        Tag = BinaryPrimitives.ReadUInt16LittleEndian(data.Slice(5, 2));
        int offset = NinePConstants.HeaderSize;
        Fid = BinaryPrimitives.ReadUInt32LittleEndian(data.Slice(offset, 4)); offset += 4;
        Name = data.ReadString(ref offset);
        AttrSize = BinaryPrimitives.ReadUInt64LittleEndian(data.Slice(offset, 8)); offset += 8;
        Flags = BinaryPrimitives.ReadUInt32LittleEndian(data.Slice(offset, 4));
    }

    public void WriteTo(Span<byte> span)
    {
        span.WriteHeaders(Size, Tag, MessageTypes.Txattrcreate);
        int offset = NinePConstants.HeaderSize;
        BinaryPrimitives.WriteUInt32LittleEndian(span.Slice(offset, 4), Fid); offset += 4;
        span.WriteString(Name, ref offset);
        BinaryPrimitives.WriteUInt64LittleEndian(span.Slice(offset, 8), AttrSize); offset += 8;
        BinaryPrimitives.WriteUInt32LittleEndian(span.Slice(offset, 4), Flags);
    }
}
